### PR TITLE
Download models in the background

### LINKED ISSUE #1260: Download Fuel models on the background
When opening a world with models that have not been downloaded, there can be a long delay before the GUI opens.
It's also not possible to close gazebo while it's doing the initial download of models.

## Desired behavior

While models are loading, the user should be able to interact with the window and close it if wanted.

## Alternatives considered

I guess we could not open the window until the models are loaded, but that may be more confusing for users.

## Implementation suggestion

* Download models in a separate thread
* Display models as they're downloaded
* Cancel loading if a signal is received
* Display status to user of what models are being loaded and what are missing
* Once we have a splash screen (https://github.com/ignitionrobotics/ign-gui/issues/336), we can use it to display this information

### PR DESCRIPTION
**Pull request by nkoenig**
Signed-off-by: Nate Koenig <nate@openrobotics.org>

# 🎉 New feature

Closes #1260.

This is an alternative to #1484.

## Summary

The PR changes the Server load to:

1. Disable Fuel downloads.
2. Parse the SDF file/string.
3. Remove all models, actors, and lights from each world.
4. Create the worlds (runners).
5. Enable Fuel downloads and force simulation to be paused. User cannot unpause.
6. Start a thread that parses the SDF file again. This time models are downloaded.
7. Add the models back into the worlds.
8. Allow the pause state in simulation to change.

The above will be followed even if nothing needs to be downloaded. 

I think the biggest downside to this approach is the double parsing of SDF.

## Test it

I haven't added tests yet. Looking for general feedback first.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by ahcorde — CHANGES_REQUESTED** [review]
`WorldByName` is not defined, is there any other related PR ?

**Review by ahcorde — CHANGES_REQUESTED** [review]
The [resource spawner](https://github.com/gazebosim/gz-sim/pull/1484/files#diff-940ea8a79df9807f0f67bcf724cd1292659a09a6e7b45e7f4ca05dc0d30233bf) and the [user command](https://github.com/gazebosim/gz-sim/pull/1484/files#diff-bc368d2827ee6222fcfad0351b4a486e819510316139f87725d2dce4ff250951) can be improved downloading models in the background.

Should I open a PR for each use case ? or can we included this changes here too ?

**Comment by ahcorde** [comment]
When I restart the simulation some models disappear (even if it's not a fuel model)

background_reset

File: src/cmd/cmdsim.rb.in [hunk 1]
```diff
@@ -109,6 +109,8 @@ COMMANDS = { 'sim' =>
   "\n"\
   "  --headless-rendering         Run rendering in headless mode                   \n"\
   "\n"\
+  "  --parallel-download          Download assets in a background thread.         \n"\
```
Comment (ahcorde) [inline src/cmd/cmdsim.rb.in:112]: I'm getting this error

```
Traceback (most recent call last):
	2: from /home/ahcorde/gz_garden/install/bin/gz:308:in `<main>'
	1: from /home/ahcorde/gz_garden/install/lib/ruby/gz/cmdsim7.rb:343:in `execute'
/home/ahcorde/gz_garden/install/lib/ruby/gz/cmdsim7.rb:329:in `parse': invalid option: --parallel-download (OptionParser::InvalidOption)
```

```suggestion
  "  --wait-for-assets          Download assets in a background thread.         \n"\
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
When I try to run with the flags I got some crashes and some error messages:

```
[GUI] [Err] [SceneManager.cc:207] Visual: [ground_plane] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [box] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [cylinder] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [sphere] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [capsule] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [ellipsoid] already exists
[GUI] [Err] [BaseStorage.hh:927] Another item already exists with name: sun
````

**Comment by nkoenig** [comment]
> `WorldByName` is not defined, is there any other related PR ?

Here it is: https://github.com/gazebosim/sdformat/pull/1121

**Comment by nkoenig** [comment]
> When I restart the simulation some models disappear (even if it's not a fuel model)
> 
> background_reset [ background_reset ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif) [ ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif)

Does this still happen to you? I've made some changes. If so, what command did you run on the command line?

**Comment by ahcorde** [comment]
> > When I restart the simulation some models disappear (even if it's not a fuel model)
> > background_reset
> >       
> >         [
> >           
> >             background_reset
> >           
> >         ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif)
> >         
> >        [ background_reset ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif)
> >       
> >         [
> >           
> >             background_reset
> >           
> >         ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif)
> >         
> >        [ ](https://user-images.githubusercontent.com/1933907/187023007-57048dbe-2450-4031-ad96-c364ba218578.gif)
> 
> Does this still happen to you? I've made some changes. If so, what command did you run on the command line?

@nkoenig I'm getting a crash, I can see the follwoing messages:

```bash
Msg [NameManager::issueNewName] (World::Skeleton | shapes) The name [ground_plane] is a duplicate, so it has been renamed to [ground_plane(1)]
Msg [NameManager::issueNewName] (World::Skeleton | shapes) The name [box] is a duplicate, so it has been renamed to [box(1)]
Msg [NameManager::issueNewName] (World::Skeleton | shapes) The name [cylinder] is a duplicate, so it has been renamed to [cylinder(1)]
Msg [NameManager::issueNewName] (World::Skeleton | shapes) The name [sphere] is a duplicate, so it has been renamed to [sphere(1)]
Msg [NameManager::issueNewName] (World::Skeleton | shapes) The name [capsule] is a duplicate, so it has been renamed to [capsule(1)]
Msg [NameManager::issueNewName] (World::Skeleton | shapes) The name [ellipsoid] is a duplicate, so it has been renamed to [ellipsoid(1)]

...

[GUI] [Err] [SceneManager.cc:207] Visual: [ground_plane] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [box] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [cylinder] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [sphere] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [capsule] already exists
[GUI] [Err] [SceneManager.cc:207] Visual: [ellipsoid] already exists
[GUI] [Err] [BaseStorage.hh:927] Another item already exists with name: sun
```

and this is the backtrace

```bash
Stack trace (most recent call last) in thread 230483:
#26   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
#25   Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7fec81f5c132, in clone
#24   Object "/lib/x86_64-linux-gnu/libpthread.so.0", at 0x7fec81e22608, in 
#23   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7ad069d1, in 
#22   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7ad05784, in QThread::exec()
#21   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7aecd3aa, in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>)
#20   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7af26434, in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>)
#19   Object "/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7fec78d0d4a2, in g_main_context_iteration
#18   Object "/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7fec78d0d3ff, in 
#17   Object "/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7fec78d0d17c, in g_main_context_dispatch
#16   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7af26e36, in 
#15   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7aed1487, in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*)
#14   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7aece809, in QCoreApplication::notifyInternal2(QObject*, QEvent*)
#13   Object "/usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5", at 0x7fec7a66a0ef, in QApplication::notify(QObject*, QEvent*)
#12   Object "/usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5", at 0x7fec7a660a65, in QApplicationPrivate::notify_helper(QObject*, QEvent*)
#11   Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7aefac29, in QObject::event(QEvent*)
#10   Object "/home/ahcorde/gz_garden/install/lib/gz-gui-7/plugins/libMinimalScene.so", at 0x7fec45688348, in gz::gui::plugins::RenderThread::RenderNext(gz::gui::plugins::RenderSync*)
#9    Object "/home/ahcorde/gz_garden/install/lib/gz-gui-7/plugins/libMinimalScene.so", at 0x7fec45695a2f, in gz::gui::plugins::RenderThreadRhiOpenGL::RenderNext(gz::gui::plugins::RenderSync*)
#8    Object "/home/ahcorde/gz_garden/install/lib/gz-gui-7/plugins/libMinimalScene.so", at 0x7fec4568b459, in gz::gui::plugins::GzRenderer::Render(gz::gui::plugins::RenderSync*)
#7    Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7aece809, in QCoreApplication::notifyInternal2(QObject*, QEvent*)
#6    Object "/usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5", at 0x7fec7a66a0ef, in QApplication::notify(QObject*, QEvent*)
#5    Object "/usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5", at 0x7fec7a660a54, in QApplicationPrivate::notify_helper(QObject*, QEvent*)
#4    Object "/usr/lib/x86_64-linux-gnu/libQt5Core.so.5", at 0x7fec7aece51a, in QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*)
#3    Object "/home/ahcorde/gz_garden/install/lib/gz-sim-7/plugins/gui/libGzSceneManager.so", at 0x7fec42fc1d6f, in gz::sim::v7::GzSceneManager::eventFilter(QObject*, QEvent*)
#2    Object "/home/ahcorde/gz_garden/install/lib/gz-sim-7/plugins/gui/libGzSceneManager.so", at 0x7fec42fc2116, in gz::sim::v7::GzSceneManagerPrivate::OnRender()
#1    Object "/home/ahcorde/gz_garden/install/lib/libgz-sim7-rendering.so.7", at 0x7fec42d3f7d4, in gz::sim::v7::RenderUtil::Update()
#0    Object "/home/ahcorde/gz_garden/install/lib/libgz-sim7-rendering.so.7", at 0x7fec42e33661, in gz::sim::v7::SceneManager::CreateLight(unsigned long, sdf::v13::Light const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long)
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
GUI is blocked when the models are being downloaded

gui_block

**Review by ahcorde — CHANGES_REQUESTED** [review]
conflicts, did you review my last comment?

**Comment by nkoenig** [comment]
> GUI is blocked when the models are being downloaded
> 
> gui_block [ gui_block ](https://user-images.githubusercontent.com/1933907/187618353-208a4e2e-c55f-4bdc-899f-0003ea0dbacd.gif) [ ](https://user-images.githubusercontent.com/1933907/187618353-208a4e2e-c55f-4bdc-899f-0003ea0dbacd.gif)

Shouldn't the GUI block if you're waiting for assets?

**Comment by nkoenig** [comment]
I unblocked the GUI when `--wait-for-assets` is true in [594a509](https://github.com/gazebosim/gz-sim/pull/1669/commits/594a50970e18e8e967fc3688ef1072095a76c0e2)

**Comment by nkoenig** [comment]
@ahcorde , how do you reset simulation?

**Comment by nkoenig** [comment]
@ahcorde , I added in your resource spawner updates in [3e3a362](https://github.com/gazebosim/gz-sim/pull/1669/commits/3e3a3629f3bc101495b46c9e4a234ad60a529894)

**Comment by nkoenig** [comment]
Reset should now work.

**Review by ahcorde — CHANGES_REQUESTED** [review]
you can also include the parallel download when a model is inserted with the command line:

Check my changes here `src/systems/user_commands/UserCommands.cc` https://github.com/gazebosim/gz-sim/pull/1484/files

**Review by ahcorde — COMMENTED** [review]
Another comment, the gui is no blocked but there is nothing in the 3D scene, 

I see two options here: 
 - add a snackbar notification
 - Add a grid, at least to understand that you might have some interactions with the 3D scene

File: src/ServerPrivate.hh [hunk 2]
```diff
@@ -100,6 +100,20 @@ namespace gz
       /// \return Path to the downloaded resource, empty on error.
       public: std::string FetchResourceUri(const common::URI &_uri);
 
+
+      /// \brief Helper function that loads an SDF root object based on
+      /// values in a ServerConfig object.
+      /// \param[in] _config Server config to read from.
+      /// \param[out] _root SDF root object.
+      /// \param[out] _outputMsgs Informational output strings.
```
Comment (ahcorde) [inline src/ServerPrivate.hh:108]:
```suggestion
      /// \param[out] _outputMsgs Informational output strings.
      /// \return
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/LevelManager.cc [hunk 3]
```diff
@@ -519,53 +518,6 @@ void LevelManager::ConfigureDefaultLevel()
   this->entityCreator->SetParent(defaultLevel, this->worldEntity);
 }
 
-/////////////////////////////////////////////////
-void LevelManager::CreatePerformers()
```
Comment (azeey) [inline src/LevelManager.cc:523]: Is this function not needed anymore?

**Review by iche033 — COMMENTED** [review]
There seems to be a GUI race condition. Sometimes I see Gazebo loads the GUI configuration from the world file correctly but sometimes it uses the default GUI config.

This is from running the `fuel_textured_mesh.sdf` world file:

GUI uses config from world file:


download_models_gui_config



GUI uses default configurations:

download_models_gui_default

**Comment by azeey** [comment]
@nkoenig since Garden will be EOL in about a month and this is a nontrivial PR, I suggest retargetting this PR to either `main` or `gz-sim8`.  Also, do we still want to get this into the Ionic release?

**Comment by nkoenig** [comment]
> @nkoenig since Garden will be EOL in about a month and this is a nontrivial PR, I suggest retargetting this PR to either `main` or `gz-sim8`. Also, do we still want to get this into the Ionic release?

I'll retarget this to `gz-sim8`.

**Comment by nkoenig** [comment]
I will open a new PR to gz-sim8 that will replace this PR.

**Pull request closed without merging**